Backport LAPACK 3.6 support

nalimilan opened the pull request:
Fedora started shipping OpenBLAS 0.2.16 without deprecated LAPACK symbols, which made the Julia build fail. So I figured it would be a good idea to backport this. I've also included the fix to check for NaNs, which looks really like a bug fix.

Cc: @andreasnoack

andreasnoack committed:
Check for NaNs in before LAPACK calls that balance the matrix. These
routines don't support NaNs and will print errors and have caused
memory errors in Julia.

Fixes #13976, #14065, #14220

Backport of aee907caf9d085d4ec9359a3c16f08edcbaece57.

andreasnoack commented:
I think some packages use the `blasfunc` function so this might actually be breaking. I bit unfortunate that I didn't split that into two commits. Maybe it's better to prepare a custom commit for 0.4 with only the `xggsvd3` part of the commit.

nalimilan commented:
Indeed. Let me have a look.

nalimilan committed:
Add xggsvd3 wrappers.

The subroutines replace xggsvd in LAPACK 3.6.0 so calls are made conditionally
on the existence of the symbols by defining a constant with the LAPACK version number.

This is a partial backport of 15c70cb.

nalimilan commented:
Wasn't very hard actually. Let's see whether the tests pass or whether I forgot to include some changes in the process.

FWIW, OpenBLAS will stop removing deprecated symbols by default in the next release: https://github.com/xianyi/OpenBLAS/issues/807 But I think it's still worth backporting.

nalimilan commented:
@tkelman Does backporting this sound OK to you?

tkelman commented:
yeah, though I just tagged so I hope this isn't urgent.Can the fedora packaging be fixed to put the deprecated functions back?

nalimilan commented:
Yeah, don't worry. I can apply it manually if needed, but I think the OpenBLAS package should be fixed to keep deprecated symbols, as it will likely break custom programs (this is actually a severe ABI break).

nalimilan commented:
@tkelman Totally unrelated, but if I were you I would change the 0.4.4 tag to point to 0.4.5. Since the release hasn't be announced, I don't think this would be a problem. It will be less confusing that suddenly skipping a version number.

tkelman commented:
Not really skipping, and I'm not force pushing anything to the release branch. 0.4.4 is ready and uploaded, but so is 0.4.5 a day later. Just need to update the links and send out the announcement.

nalimilan commented:
No need to force-push commits to remove a tag. Changing tags is much less disruptive than messing with commits history.

tkelman commented:
Why would I use tags that don't match the contents of VERSION? That seems more confusing to me than "0.4.4 exists but has a known bug, you should use 0.4.5 which is otherwise effectively the same."

nalimilan commented:
Of course, I would also have reverted the version to 0.4.4. Anyway, I don't care that much.

tkelman commented:
version not being monotonic could cause issues...
